### ISSUE TITLE
profiling tools for source editor 

### ISSUE THREAD

**Issue by ian29** [mapbox/mapbox-studio-classic#146]
**tl;dr** TM2 should help users know how dangerous / unreasonable their potential vector tiles are. given [profiling tools in the style editor](https://github.com/mapbox/tm2/issues/20#issuecomment-17792932), ideally TM2 provides per-zoom-level stats. 

My first escapade into rendering + _optimizing_ vector tiles left me wishing i knew more about how impactful my queries were - felt like i was writing optimization queries without knowing how much they optmized anything. i dont know how feasible this is(nt), but i found myself rendering test mbtiles and wanting/generating stats like these below. 

Render-time profiling like in the style editor doesnt make sense so much as max, avg tile size, and zoom level "size", so I dont think copying over code from the style editor makes sense. 

```
select 
    zoom_level,
    count(zoom_level),
    max(length(tile_data)) / 1000,
    avg(length(tile_data)) / 1000,
    sum(length(tile_data)) / 1000
from tiles
group by zoom_level;
```

| z | tiles | max tile size (KB) | avg tile size (KB) | z level size (KB) |
| --- | --- | --- | --- | --- |
| 0 | 1 | 10392 | 10392.539 | 10392 |
| 1 | 1 | 10487 | 10487.1 | 10487 |
| 2 | 4 | 10579 | 6702.82575 | 26811 |
| 3 | 8 | 10608 | 5287.871625 | 42302 |
| 4 | 26 | 8154 | 1628.54 | 42342 |
| 5 | 60 | 6316 | 698.407833333334 | 41904 |
| 6 | 138 | 2714 | 305.877108695652 | 42211 |
| 7 | 349 | 1520 | 122.68358739255 | 42816 |
| 8 | 948 | 919 | 45.9349683544304 | 43546 |
| 9 | 2849 | 624 | 15.5313734643735 | 44248 |
| 10 | 8835 | 459 | 5.22136570458404 | 46130 |
| 11 | 25581 | 280 | 1.9388886282788 | 49598 |
| 12 | 64298 | 176 | 0.861049830476842 | 55363 |
| 13 | 128463 | 110 | 0.49414084211018 | 63478 |
### next actions

@yhahn if i were to poke at this on the weekends, would you recommend poking around in the style editor profiling code? or just start from scratch. 

cc/ @ajashton

**Comment by yhahn** [comment]
@ian29 hrm, you may want to start by reading this function:

https://github.com/mapbox/tilelive-vector/blob/master/index.js#L239-L340

This is built around choosing tiles that shoudl bust up render time but could easily be repurposed to focus on VT tile size.

I haven't built a UI to run/represent the stats generated here but your notes/goals are exactly what it's going for.

**Comment by rclark** [comment]
@ian29 I was working my way into this kind of thing while thinking about https://github.com/mapbox/mapbox-streets/issues/740. I was thinking of ways to take it a step deeper and be able to give some kinds of statistics about geometries within the tile -- how many coordinates? ave distance between coordinates? number of repeated coordinates? If you have anything started around this lets sync up

**Comment by rclark** [comment]
What I'd like to see are a few different ways to "profile"
- Give me stats about one `[z, x ,y]`
- Give me min/max/ave stats about these `[ [z1, x1, y1], [z2, x2, y2], ... ]`
- Give me stats about the tiles that I see when I move around and look at different parts of the map

As I mentioned before, I'd like to not only see tile sizes, but also geometry details. I think these kinds of parameters are important to try and understand what further optimizations are worth pursuing. Ideally we the system would allow for "pluggable" functions that track some particular statistic.

#### Potential approaches
- Right now I've developed a separate standalone app for gathering / reducing tile statistics. Hacked a little into TM2 so it sends to the other app _every single vector tile that it renders_. Separate app is responsible for data reduction and visualization.
- Alternatively, build UI elements into TM2 that give you some sense of on-the-fly optimization. I'm seeing a page that lets you a) click a tile on the map, get stats, b) give some parameters to have it gather stats on some set of tiles, c) turn on a "logger" or something that will track stats on tiles as you browse the map.

Curious @ian29 and @yhahn how deep you think its worth diving here. I think its pretty important; gathering more of this kind of information could help us make decisions about how to automatically optimize vector tiles without user input. Ideally TM2 would not require that users "know how dangerous / unreasonable their potential vector tiles are". These kinds of optimizations are totally orthogonal to the task that users are actually trying to accomplish -- I want to make a map that portrays some data the way I want it portrayed. I think we should try to make vector tiles an invisible part of the rendering pipeline that users almost never have to think about.

**Comment by e-n-f** [comment]
I don't know if this goes along with what you're doing @rclark or if it's completely orthogonal to it, but my hope is still to finish making an alternate vector tile producing tool where you can throw in an arbitrary shapefile or GeoJSON and get back a set of vectors that Just Work, automatically simplified to throw away as much detail as is necessary at low zooms and while still being complete and detailed at high zooms. Trying to do manual simplifications for low zoom levels seems like it would always be way too frustrating otherwise.

Ideally in my mind this would even be part of the Mapbox service so you just get a map ID back instead of ever having to even upload a tileset from your local computer.

**Comment by rclark** [comment]
Starting to flesh out some ideas based on our discussion. The blueprint is
- branch of `tilelive-vector` where you can [set a flag](https://github.com/mapbox/tilelive-vector/blob/more-stats/index.js#L184) that will perform some [more detailed statistics gathering](https://github.com/mapbox/tilelive-vector/blob/more-stats/tile-profiler.js)
- at first, [flip the flag](https://github.com/mapbox/tm2/blob/more-stats/index.js#L232) in tm2 so that it will gather more detailed stats, and then [adjust the stats gatherer](https://github.com/mapbox/tm2/blob/more-stats/lib/style.js#L351-L368) so that we can collect / reduce more complex info. For example before we could only collect min/max/avg values _by zoom level_, now we could collect min/max/avg values _by layer, by zoom level_, or any arbitrary depth of indexing. This could also allow statistics to get collected across changes that you might make in trying to optimize.

Next up will be trying it out for a little while, and then experimenting with 
- testing (tm2 doesn't work in my current, very restricted network environment)
- think a little about UI for things like turing detailed gathering on/off, perhaps collecting different stat "sets" so you can compare stats before and after you make a change in an attempt to optimize tiles

**Comment by rclark** [comment]
Tested this approach out and it appears to be working great. My goal is to make it possible for a user to collect named "stat sets". User experience would eventually include changes to this piece of the interface:

![screen shot 2014-03-07 at 10 11 25 am](https://f.cloud.github.com/assets/515424/2359753/b410c72c-a61b-11e3-8951-7b15e2dfbcf8.png)

User might be click a button, assign a name for a stat set, then statistics are collected within that set until user clicks stop. User should be able to toggle between different stats and compare stat sets.

The next step then is to make some adjustments to the [stats content that lives in the cookie]() to make it possible to track and distinguish between different stat sets.

**Comment by yhahn** [comment]
Rock. Is the cookie-as-parallel-to-tile-transport mechanism scaling ok for you so far?

**Comment by rclark** [comment]
Not sure yet about scaling. I'm about to try, but in the event it doesn't scale, we might be able to reduce the cookie just a storing stat-set-names and then writing an API call the client can use to get stat values.

**Comment by rclark** [comment]
@yhahn I'm actually not seeing cookies getting truncated, so I think we're still alright.

@ian29 there are some adjustments in the `more-stats` branch I'd be interested in your feedback on. Specifically, if you browse around your map and then check out `http://localhost:3000/style/statistics.json?id=your-style-id` you ought to get a JSON object that you could reduce in order to create tables like the one you've outlined above.

If you're interested in seeing some geometry statistics you need to change `package.json` to use the `more-stats` branch of tilelive-vector. Then the statistics object should include more info.

Perhaps the next step would be better UI integration (bouncing between different types of statistics, recording and comparing "stat sets", etc). I made some stabs at this but it looked like it would get messy really fast. I would want to bounce some ideas off of people before diving deeper in that direction.

**Comment by samanpwbb** [comment]
@yhahn did this:
![screen shot 2014-09-15 at 1 59 02 pm](https://cloud.githubusercontent.com/assets/108094/4276052/ffb16eac-3d01-11e4-9091-c06522406aad.png)

**Comment by ian29** [comment]
hooray its back! cc @karenzshea
